### PR TITLE
core: init native interops in the genesis block

### LINKED ISSUE #836: Add native interops initialization into the genesis block
Exactly the same way NEO 3.0 does, but without removing old transactions yet. Depends on #779.

### PR DESCRIPTION
**Pull request by AnnaShaleva**
closes #836

### CODE REVIEW

File: pkg/core/util.go [hunk 1]
```diff
@@ -96,6 +99,29 @@ func createGenesisBlock(cfg config.ProtocolConfiguration) (*block.Block, error)
 	return b, nil
 }
 
+func deployNativeContracts() *transaction.Transaction {
+	buf := io.NewBufBinWriter()
+	emit.Syscall(buf.BinWriter, "Neo.Native.Deploy")
+	// script := buf.Bytes()
+	return &transaction.Transaction{
+		Version:         0,
```
Comment (fyrchik) [inline pkg/core/util.go:107]: There is no need to specify `0` explicitly, it is the default value.

**Comment by roman-khimov** [comment]
@fyrchik: please take a look, I'll rebase it afterwards.

**Comment by roman-khimov** [comment]
Updated and improved, it's important to merge this tomorrow, so have a look at it.

File: pkg/core/native/native_neo.go [hunk 2]
```diff
@@ -24,10 +23,37 @@ type NEO struct {
 	GAS *GAS
 }
 
-const neoSyscallName = "Neo.Native.Tokens.NEO"
+// keyWithVotes is a serialized key with votes balance. It's not deserialized
+// because some uses of it imply serialized-only usage and converting to
```
Comment (AnnaShaleva) [inline pkg/core/native/native_neo.go:27]: Typo `imply`

File: pkg/core/state/native_state.go [hunk 3]
```diff
@@ -32,6 +57,31 @@ func (s *NEP5BalanceState) DecodeBinary(r *io.BinReader) {
...
 
+// NEOBalanceStateFromBytes converts serialized NEOBalanceState to structure.
+func NEOBalanceStateFromBytes(b []byte) (*NEOBalanceState, error) {
+	balance := new(NEOBalanceState)
+	if len(b) == 0 {
+		return balance, nil
+	}
+	r := io.NewBinReaderFromBuf(b)
+	balance.DecodeBinary(r)
+
```
Comment (fyrchik) [inline pkg/core/state/native_state.go:70]: Why do you have an empty line here but not in `Bytes()` ?

File: pkg/core/dao/dao.go [hunk 4]
```diff
@@ -471,9 +472,19 @@ func (dao *Simple) DeleteStorageItem(scripthash util.Uint160, key []byte) error
 
 // GetStorageItems returns all storage items for a given scripthash.
 func (dao *Simple) GetStorageItems(hash util.Uint160) (map[string]*state.StorageItem, error) {
+	return dao.GetStorageItemsWithPrefix(hash, nil)
+}
+
+// GetStorageItemsWithPrefix returns all storage items with given prefix for a
+// given scripthash.
+func (dao *Simple) GetStorageItemsWithPrefix(hash util.Uint160, prefix []byte) (map[string]*state.StorageItem, error) {
```
Comment (fyrchik) [inline pkg/core/dao/dao.go:365]: I think it is better not to return map in DAO as the returned order is unpredictable (remember #822).
And if we need random access, new method or a use of bigger prefix are OK IMO.

File: pkg/core/native/native_neo.go [hunk 5]
```diff
@@ -167,6 +167,27 @@ func (n *NEO) increaseBalance(ic *interop.Context, h util.Uint160, si *state.Sto
 	if amount.Sign() == 0 {
 		return nil
 	}
+	oldAcc, err := ic.DAO.GetAccountState(h)
+	if err != nil {
+		return err
+	}
+	if err := n.ModifyAccountVotes(oldAcc, ic.DAO, (&big.Int{}).Not(&acc.Balance)); err != nil {
```
Comment (fyrchik) [inline pkg/core/native/native_neo.go]: It is not `Not`, but `Neg`.
We also use `new(big.Int)` in every other place.

File: pkg/consensus/consensus.go [hunk 6]
```diff
@@ -479,22 +479,12 @@ func (s *service) getVerifiedTx(count int) []block.Transaction {
 	return res
 }
 
-func (s *service) getValidators(txx ...block.Transaction) []crypto.PublicKey {
+func (s *service) getValidators(_ ...block.Transaction) []crypto.PublicKey {
 	var (
 		pKeys []*keys.PublicKey
 		err   error
 	)
-	if len(txx) == 0 {
```
Comment (fyrchik) [inline pkg/consensus/consensus.go:487]: When invoked with non-zero sized slice, it is used for calculating `NextConsensus` field.
DBFT needs to be changed, but for now we can return `GetNextBlockValidators` when invoked with non-zero slice.
https://github.com/neo-project/neo/blob/master/src/neo/Consensus/ConsensusContext.cs#L349

File: pkg/core/native/native_neo.go [hunk 7]
```diff
@@ -446,7 +450,7 @@ func (n *NEO) getNextBlockValidators(ic *interop.Context, _ []vm.StackItem) vm.S
 func (n *NEO) GetNextBlockValidatorsInternal(bc blockchainer.Blockchainer, d dao.DAO) (keys.PublicKeys, error) {
 	si := d.GetStorageItem(n.Hash, nextValidatorsKey)
 	if si == nil {
-		return bc.GetStandByValidators()
```
Comment (fyrchik) [inline pkg/core/native/native_neo.go]: If `validatorsCount` is not yet initialized, `GetValidatorsInternal` can fail, no?
https://github.com/neo-project/neo/blob/master/src/neo/SmartContract/Native/Tokens/NeoToken.cs#L264

File: pkg/consensus/consensus.go (hunk 6 above)
Comment (roman-khimov) [reply pkg/consensus/consensus.go:487]: I don't think DBFT needs any change wrt this. We already do return next block validators from `GetValidators`.

File: pkg/core/native/native_neo.go (hunk 7 above)
Comment (roman-khimov) [reply pkg/core/native/native_neo.go]: It's initialized now in `Initialize` at block 0, so I don't think we have any problem.

File: pkg/core/native/native_neo.go (hunk 5 above)
Comment (roman-khimov) [reply pkg/core/native/native_neo.go]: > It is not Not, but Neg.

Yep, it was fixed in some subsequent commit, maybe I need to shuffle these changes a little.

File: pkg/core/dao/dao.go (hunk 4 above)
Comment (roman-khimov) [reply pkg/core/dao/dao.go:365]: Argh, I keep forgetting about this. But then we probably also don't need `string` conversion here.

Comment (roman-khimov) [reply pkg/core/dao/dao.go:365]: Well, that may be a separate task as our current `GetStorageItems` API does the same thing (and that's why `WithPrefix` inherited it).

Comment (roman-khimov) [reply pkg/core/dao/dao.go:365]: #909.

**Comment by roman-khimov** [comment]
Updated to mention #865 as it technically fixes it now.

**Pull request merged**